### PR TITLE
Add support for bigcache and freecache algorithms

### PR DESCRIPTION
**Pull request by mangalaman93**

### CODE REVIEW

**Comment by mangalaman93** [comment]
[manishrjain](/manishrjain) please have a look.

**Comment by dgryski** [comment]
I'm happy to merge this but I'll wait on a LGTM from @manishrjain

**Comment by dgryski** [comment]
One thing that will affect the timings in the "real world" is the requirement to serialize to`[]byte` instead of being able to store `interface{}` as values.  Not sure if this matters to you.

**Comment by mangalaman93** [comment]
We are looking at caches for speeding up responses from [badger](https://github.com/dgraph-io/badger). Badger stores sequence of bytes, hence, this is what we want.

I was wondering, though, what traces did you use to compare these cache algorithms?

**Comment by dgryski** [comment]
A set of domain names for email addresses from a leaked user database from a few years ago.  It would be interesting to see comparative graphs for the different algorithms at different Zipf distributions (which will be easy now that we have a generator in the repo)

**Comment by mangalaman93** [comment]
Check this out https://docs.google.com/spreadsheets/d/1c5AbsHPh-z6KyxkiBJ4ipNR7U7QVuSmR38fym8NeC50/edit#gid=1156181048

**Comment by dgryski** [comment]
Nice.  Also interesting would be the actual time taken to actually process the trace.  ARC tends to be slower (due to all the pointer chasing) even though it has a high hit rate.  Tiny-lfu and Clock-pro tend to be much faster in addition to being non-patent-encumbered :)

**Comment by dgryski** [comment]
In terms of cache hit ratios, both `bigcache` and `freecache` appear to be LRU-based.  Thus, I would expect their hit ration to mirror LRU, but their memory usage and/or GC impact to be lower than the linked-list-based implementation current included in the benchmark set.

**Comment by manishrjain** [comment]
> In terms of cache hit ratios, both `bigcache` and `freecache` appear to be LRU-based

From @mangalaman93 's understanding, bigcache is largely a FIFO queue, not an LRU cache. We see their hit ratios getting bad as the number of keys exceeds what can fit in the cache.

**Comment by dgryski** [comment]
That's entirely possible -- my analysis was based on ~30s of glancing at the repos :)

File: cachetest/main.go [hunk 1]
```diff
@@ -81,6 +83,55 @@ func main() {
 
 	switch *alg {
 
+	case "freecache":
+
+		// HDR (24) + Key (8) + Value (8) = 40
+		cache := freecache.NewCache(*n * 40)
+		f = func(s string) bool {
+			if i, err := cache.Get([]byte(s)); err == freecache.ErrNotFound {
```
Comment (manishrjain) [inline cachetest/main.go]: I'd do the conversion to []byte once, and use it in multiple places.

**Comment by mangalaman93** [comment]
@dgryski ready to merge!

**Comment by mangalaman93** [comment]
Thank you so much!

**Pull request merged**